### PR TITLE
[fix] [broker] Update topic policies as much as possible when some ex was thrown

### PR DESCRIPTION
**Pull request by poorbarcode**
### Motivation

After the topic policies update, there are many components will be updated one by one, even if the config of components has not been modified. There are the 11 components that need update:
- `7` rate limiters(`publish`, `dispatch topic-level`, `dispatch subscription-level`,  `dispatch resourceGroup-level`, `subscribe API`, `replication`, `shadow topic replication`)
- update ManagedLedger configs(`retention`, `offloader`)
- start/stop replication
- start/stop compaction
- start/stop deduplication

Once a component update fails, the following update will be skipped. It would cause a confusing thing: you want to set a retention policy, but it will be skipped due to the `update subscribe rate limiter` failure (you did not edit the `subscribe rate limitation policy`)

Since none of the components in the above list have any additional dependencies for individual updates, ensuring success as much as possible is appropriate.

### Modifications
- Update topic policies as much as possible even if some component updates fail, all component updates are still in the same thread, and they still update one by one, just throw the error later.
- Rename `updatePublishDispatcher` to `updatePublishRateLimiter`

### Documentation

- [ ] `doc` 
- [ ] `doc-required` 
- [x] `doc-not-needed` 
- [ ] `doc-complete` 

### Matching PR in forked repository

PR in forked repository: x

### CODE REVIEW

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/admin/TopicPoliciesTest.java [hunk 1]
```diff
@@ -3157,4 +3163,49 @@ public void testProduceChangesWithEncryptionRequired() throws Exception {
                 });
     }
 
+    @Test
+    public void testUpdateRetentionWhenOtherComponentUpdateFail() throws Exception {
```
Comment (codelipenghui) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/admin/TopicPoliciesTest.java:3167]:
```suggestion
    public void testUpdateRetentionWithPartialFailure() throws Exception {
```

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java [hunk 2]
```diff
@@ -3061,39 +3062,67 @@ public CompletableFuture<Void> onPoliciesUpdate(@Nonnull Policies data) {
...
+            .exceptionally(ex -> {
+                log.error("[{}] update namespace polices : {} error", this.getName(), data, ex);
+                throw FutureUtil.wrapToCompletionException(ex);
+            });
+    }
 
-        List<CompletableFuture<Void>> producerCheckFutures = new ArrayList<>(producers.size());
-        producers.values().forEach(producer -> producerCheckFutures.add(
-                producer.checkPermissionsAsync().thenRun(producer::checkEncryption)));
+    private List<CompletableFuture<Void>> updateComponentPolicies(Policies namespacePolicies) {
```
[the remark this answers is not in the document]
Comment (poorbarcode) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java]: Renamed

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/admin/TopicPoliciesTest.java (hunk 1 above)
Comment (poorbarcode) [reply pulsar-broker/src/test/java/org/apache/pulsar/broker/admin/TopicPoliciesTest.java]: Renamed

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java [hunk 3]
```diff
@@ -3061,39 +3062,67 @@ public CompletableFuture<Void> onPoliciesUpdate(@Nonnull Policies data) {
...
+        updateComponentsFutureList.add(CompletableFuture.runAsync(
+                () -> updateDispatchRateLimiter(), MoreExecutors.directExecutor()));
+        updateComponentsFutureList.add(CompletableFuture.runAsync(
+                () -> updateSubscribeRateLimiter(), MoreExecutors.directExecutor()));
+        updateComponentsFutureList.add(CompletableFuture.runAsync(
+                () -> updatePublishRateLimiter(), MoreExecutors.directExecutor()));
+        if (namespacePolicies != null) {
+            updateComponentsFutureList.add(CompletableFuture.runAsync(
+                    () -> updateResourceGroupLimiter(namespacePolicies), MoreExecutors.directExecutor()));
+        }
```
[the remark this answers is not in the document]
Comment (poorbarcode) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java]: Sure, added a new field `lastUpdatedNamespacePolicies` to record the latest updated namespace-level policies.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/AbstractTopic.java [hunk 4]
```diff
@@ -158,6 +158,8 @@ public abstract class AbstractTopic implements Topic, TopicPolicyListener<TopicP
     protected volatile boolean transferring = false;
     private volatile List<PublishRateLimiter> activeRateLimiters;
 
+    protected Policies lastUpdatedNamespacePolicies;
```
Comment (codelipenghui) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/AbstractTopic.java:161]: Oh, sorry. I thought it was able to get from somewhere before.
Pass via the params is better. :)

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java [hunk 5]
```diff
@@ -3061,39 +3064,68 @@ public CompletableFuture<Void> onPoliciesUpdate(@Nonnull Policies data) {
...
 
-        updateDispatchRateLimiter();
-
-        updateSubscribeRateLimiter();
-
-        updatePublishDispatcher();
-
-        updateResourceGroupLimiter(data);
+    private List<CompletableFuture<Void>> applyUpdatedPolicies() {
+        List<CompletableFuture<Void>> updateComponentsFutureList = new ArrayList<>();
```
Comment (codelipenghui) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java:3086]:
```suggestion
        List<CompletableFuture<Void>> applyPoliciesFutureList = new ArrayList<>();
```

**Review by codelipenghui — APPROVED** [review]

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (codelipenghui) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java]: Oh, sorry. I thought it was able to get from somewhere before.
Pass via the params is better. :)

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java (hunk 5 above)
Comment (poorbarcode) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java]: Renamed

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/AbstractTopic.java (hunk 4 above)
Comment (poorbarcode) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/AbstractTopic.java]: Sure, reverted

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (poorbarcode) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java]: > Oh, sorry. I thought it was able to get from somewhere before.
Pass via the params is better. :)

Sure, reverted

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java [hunk 6]
```diff
@@ -3061,39 +3062,67 @@ public CompletableFuture<Void> onPoliciesUpdate(@Nonnull Policies data) {
...
+        applyPoliciesFutureList.add(CompletableFuture.runAsync(
+                () -> checkReplicationAndRetryOnFailure(), MoreExecutors.directExecutor()));
+        applyPoliciesFutureList.add(CompletableFuture.runAsync(
+                () -> checkDeduplicationStatus(), MoreExecutors.directExecutor()));
+        applyPoliciesFutureList.add(CompletableFuture.runAsync(
+                () -> checkPersistencePolicies(), MoreExecutors.directExecutor()));
+        applyPoliciesFutureList.add(CompletableFuture.runAsync(
+                () -> preCreateSubscriptionForCompactionIfNeeded(), MoreExecutors.directExecutor()));
+
+        return applyPoliciesFutureList;
```
Comment (Technoboy-) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java:3125]: Add a new private method :
```
private CompletableFuture<Void> runDirectly(Runnable runnable) {
        return CompletableFuture.runAsync(runnable, MoreExecutors.directExecutor());
    }
```

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java [hunk 7]
```diff
@@ -3736,40 +3765,30 @@ public void onUpdate(TopicPolicies policies) {
...
-        .thenCompose(__ -> checkPersistencePolicies())
-        .thenAccept(__ -> log.info("[{}] Policies updated successfully", topic))
-        .exceptionally(e -> {
-            Throwable t = FutureUtil.unwrapCompletionException(e);
-            log.error("[{}] update topic policy error: {}", topic, t.getMessage(), t);
-            return null;
-        });
+
+        // Apply policies for components.
+        FutureUtil.waitForAll(applyUpdatedPolicies(null))
```
Comment (Technoboy-) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java:3776]: `null` param is confuse. Better to add a new method like `applyPolicies`

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java [hunk 8]
```diff
@@ -3736,40 +3765,30 @@ public void onUpdate(TopicPolicies policies) {
...
-                if (dispatcher != null) {
-                    dispatcher.updateRateLimiter();
-                }
-            }));
+            Dispatcher dispatcher = sub.getDispatcher();
+            if (dispatcher != null) {
+                dispatcher.updateRateLimiter();
+            }
         });
         return FutureUtil.waitForAll(subscriptionCheckFutures);
```
Comment (Technoboy-) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java:3793]: Seem not add future to  `subscriptionCheckFutures`

Comment (poorbarcode) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java]: Fixed

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java (hunk 6 above)
Comment (poorbarcode) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java:3116]: Sure, added a tool method in `FutureUtil`

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java (hunk 7 above)
Comment (poorbarcode) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java]: Sure, split to 2 methods: `applyUpdatedTopicPolicies` and `applyUpdatedNamespacePolicies `

**Review by Technoboy- — APPROVED** [review]

**Pull request merged**